### PR TITLE
feat(goal_planner): add extra front margin for collision check considering stopping distance

### PR DESCRIPTION
**Pull request by kosuke55**
# Description

currently, the collision check margin is the same value in any direction.
In this PR, add extra front margin dynamically considering stopping distance.

### <samp>🤖 Generated by Copilot at 84fbda8</samp>

This pull request enhances the goal planner module of the behavior path planner by using spline interpolation and extra stopping margin for pull over path planning. It adds a new parameter `object_recognition_collision_check_extra_max_stopping_margin` to control the extra margin and updates the collision check functions and the documentation accordingly. It also modifies the files `goal_planner_module.cpp`, `goal_planner.param.yaml`, `goal_planner_parameters.hpp`, `utils.hpp`, `manager.cpp`, and `utils.cpp`.

## Related links

https://github.com/autowarefoundation/autoware_launch/pull/520

## Tests performed

psim

https://github.com/autowarefoundation/autoware.universe/assets/39142679/f7536332-0e97-43d6-8ff3-6e3869003ea0

https://evaluation.tier4.jp/evaluation/reports/466af58f-2513-5a1c-9df3-efcffe9bae0e?project_id=prd_jt

## Interface changes

none

## Effects on system behavior

none

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by kosuke55** [comment]
when `object_recognition_collision_check_margin: 0.5` ego can not restart because of `surround_obstacle_checker`

```
surround_check_distance: 0.5 # if objects exist in this distance, transit to "exist-surrounding-obstacle" status [m]
```

image

File: planning/behavior_path_planner/src/utils/utils.cpp [hunk 1]
```diff
@@ -621,6 +621,29 @@ bool checkCollisionBetweenPathFootprintsAndObjects(
   return false;
 }
 
+bool checkCollisionWithExtraStoppingMargin(
```
Comment (kyoichi-sugahara) [inline planning/behavior_path_planner/src/utils/utils.cpp:624]: @kosuke55
could you move this util function in safety_check.cpp?

Comment (kosuke55) [reply planning/behavior_path_planner/src/utils/utils.cpp:624]: @kyoichi-sugahara OK! that's better!

File: planning/behavior_path_planner/src/utils/utils.cpp [hunk 2]
```diff
@@ -621,6 +621,29 @@ bool checkCollisionBetweenPathFootprintsAndObjects(
...
+  const PathWithLaneId & ego_path, const PredictedObjects & dynamic_objects,
+  const double base_to_front, const double base_to_rear, const double width,
+  const double maximum_deceleration, const double margin, const double max_extra_stopping_margin)
```
Comment (kyoichi-sugahara) [inline planning/behavior_path_planner/src/utils/utils.cpp:627]: how about 
`collision_distance_threshold` or `collision_distance` hmm.. something better variable name for `margin`?

**Review by kyoichi-sugahara — APPROVED** [review]
LGTM

File: planning/behavior_path_planner/src/utils/utils.cpp (hunk 2 above)
Comment (kosuke55) [reply planning/behavior_path_planner/src/utils/utils.cpp]: rename to collision_check_margin considering params

https://github.com/autowarefoundation/autoware.universe/pull/4674/commits/99229aac24bc6c2550fc888f153dca6883762947

File: planning/behavior_path_planner/src/utils/utils.cpp (hunk 1 above)
Comment (kosuke55) [reply planning/behavior_path_planner/src/utils/utils.cpp]: moved in https://github.com/autowarefoundation/autoware.universe/pull/4674/commits/accb2dfd9bfd1c4be772e6b069e5e9a970ee4660

**Pull request merged**